python/ray/data/datasource/file_based_datasource.py

```diff
@@ -792,3 +774,37 @@ def _resolve_kwargs(
         kwarg_overrides = kwargs_fn()
         kwargs.update(kwarg_overrides)
     return kwargs
+
+
+Uri = TypeVar("Uri")
+Meta = TypeVar("Meta")
+
+
+def _fetch_metadata_parallel(
+    uris: List[Uri],
+    fetch_func: Callable[[List[Uri]], List[Meta]],
+    desired_uris_per_task: int,
+    **ray_remote_args,
+) -> Iterator[Meta]:
+    """Fetch file metadata in parallel using Ray tasks."""
+    num_cpus = 0.5
+    remote_fetch_func = cached_remote_fn(fetch_func, num_cpus=num_cpus)
+    if ray_remote_args:
+        remote_fetch_func = remote_fetch_func.options(**ray_remote_args)
+    available_cpus = _estimate_available_parallelism()
+    parallelism = min(
+        # Choose a parallelism that results in a # of metadata fetches per task that
+        # dominates the Ray task overhead while ensuring good parallelism.
+        # Always launch at least 2 parallel fetch tasks.
+        max(len(uris) // desired_uris_per_task, 2),
+        # Oversubscribe cluster CPU by 2x since these tasks are I/O-bound.
+        round(available_cpus / num_cpus),
```

Comment:
I think we can remove line 801, since as long as the tasks are long enough it doesn't matter to have more tasks after that.

Reply:
Ah good point! I.e. we should keep the 0.5 CPUs resource request to provide some backpressure so we don't end up starting way more workers than CPUs, but ditch the truncating of parallelism by a 2x oversubscription, since we can expect tasks to queue due to the CPU-based backpressure so the scheduling/dispatch overhead should be minimal for queued tasks and therefore amortized if we have tasks >> CPUs?

Reply:
Btw, this ended up leading to pretty big slowdowns of the 10k files and 40k files cases, ~2x slower in both cases, so the task scheduling/dispatching overhead is still non-negligible upon queueing. Would you still vote for removing this?

Reply:
Interesting. I'd still vote to remove it, since we don't understand why it should be faster (40k/16 = 2500 tasks, which should be less than a second of overhead).

Reply:
Could it be using 0.5 CPUs is actually slower than 1 CPU due to worker startup time? I don't think the worker pool caches more than 1 worker per CPU.

Reply:
It actually ended up being the 0.5 CPUs vs 0.25 CPUs difference, with the latter being 2x faster. Keeping `num_cpus` fixed and removing this CPU oversubscription bound resulted in ~ the same performance.

I wouldn't expect worker startup time to be the issue here, since we're talking about queued, short tasks or large (e.g. 40 second) tasks, with CPU oversubscription in both cases. In either case, more workers would allow us to drain the queue faster.